src/auth/linking.ts

```diff
@@ -0,0 +1,242 @@
+import aes from "keystore-idb/lib/aes/index.js"
+import config from "keystore-idb/lib/config.js"
+import rsa from "keystore-idb/lib/rsa/index.js"
+import utils from "keystore-idb/lib/utils.js"
+import { KeyUse, SymmAlg, HashAlg, CharSize } from "keystore-idb/lib/types.js"
+import * as did from "../did/index.js"
+import * as ucan from "../ucan/index.js"
+import { publishOnChannel } from "./index.js"
+import * as auth from "./index.js"
+
+type LinkingRole = "CONSUMER" | "PRODUCER"
+
+type LinkingStep = "BROADCAST" | "NEGOTIATION" | "DELEGATION"
+
+type LinkingState = {
+  sessionKey: CryptoKey | null
+  temporaryRsaPair: CryptoKeyPair | null
+  role: LinkingRole | null
+  step: LinkingStep | null
+}
+
+const ls: LinkingState = {
+  sessionKey: null,
+  temporaryRsaPair: null,
+  role: null,
+  step: null
+}
+
+const resetLinkingState = () => {
+  ls.sessionKey = null
+  ls.temporaryRsaPair = null
+}
+
+const nextStep = () => {
+  switch (ls.step) {
+    case "BROADCAST":
+      ls.step = "NEGOTIATION"
+      break
+    case "NEGOTIATION":
+      ls.step = "DELEGATION"
+      break
+    default:
+      ls.step = "BROADCAST"
+  }
+}
+
+export const startLinkingConsumer = async (username: string): Promise<null> => {
+  ls.role = "CONSUMER"
+  ls.step = "BROADCAST"
+  await auth.openChannel(username)
+  await sendTemporaryExchangeKey()
+  return null
+}
+
+export const startLinkingProducer = async (username: string): Promise<null> => {
+  ls.role = "PRODUCER"
+  ls.step = "BROADCAST"
+  await auth.openChannel(username)
+  return null
+}
+
+export const handleMessage = async (event: MessageEvent): Promise<any> => {
+  const { data } = event
+  const message = new TextDecoder().decode(data.arrayBuffer ? await data.arrayBuffer() : data)
+  console.debug("Linking Status", ls)
+  console.debug("message (raw)", message)
+  switch (ls.role) {
+    case "CONSUMER":
+      await handleConsumerMessage(message)
+      break
+    case "PRODUCER":
+      await handleProducerMessage(message)
+      break
```

Comment:
I don't know if this makes this more readable or not...

Reply:
If it were me, I'd break these into separate modules. e.g. `Producer.handleMessage`

Reply:
i.e. I'm also not sold on this being the right abstraction

Reply:
I'm certainly not opposed to that... would need to sort through the impact on the channel abstraction, but it's probably safe to assume that opening a channel will always require some level of callback for received messages (although the signature will change, which I was trying to keep consolidated in WSS implementation).

Anyway, this is 100% the layer where I was still noodling on what makes the most sense for potential future usage.